### PR TITLE
git: don't update submodules in offline mode

### PR DESCRIPTION
**Pull request by brgl**
When we're running in --offline mode, don't try to update git submodules or else we're bail out with a network error if it's actually inaccessible.

This addresses the issue raised here: https://github.com/rust-lang/cargo/issues/10727

### CODE REVIEW

**Comment by ehuss** [comment]
Can you add a test that demonstrates the issue?  From your response on the issue, I still can't tell when this would be needed.

**Comment by ehuss** [comment]
There are several tests in https://github.com/rust-lang/cargo/blob/master/tests/testsuite/git.rs that set up git submodules as dependencies.

**Comment by brgl** [comment]
Building current cargo master fails for me:

```
$ cargo build --locked
   Compiling kstring v2.0.0
   Compiling openssl-sys v0.9.74
   Compiling libz-sys v1.1.8
   Compiling libnghttp2-sys v0.1.7+1.45.0
   Compiling serde v1.0.137
   Compiling ignore v0.4.18
   Compiling im-rc v15.1.0
   Compiling strip-ansi-escapes v0.1.1
   Compiling flate2 v1.0.24
   Compiling libssh2-sys v0.2.23
   Compiling openssl v0.10.40
error: `MaybeUninit::<T>::assume_init` is not yet stable as a const fn
   --> /home/brgl/.cargo/registry/src/github.com-1ecc6299db9ec823/kstring-2.0.0/src/string.rs:844:17
    |
844 |                 std::mem::MaybeUninit::uninit().assume_init()
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `kstring` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
```

Is there any fix for that yet?

**Comment by epage** [comment]
@brgl you need to be building with 1.59

**Comment by Eh2406** [comment]
What version of Rust are you using? [assume_init](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.assume_init) was stabilized for const fn in 1.59.

File: src/cargo/sources/git/utils.rs [hunk 1]
```diff
@@ -177,7 +177,9 @@ impl GitDatabase {
             Some(c) => c,
             None => GitCheckout::clone_into(dest, self, rev, cargo_config)?,
         };
-        checkout.update_submodules(cargo_config)?;
+        if !cargo_config.offline() {
+            checkout.update_submodules(cargo_config)?;
+        }
```
Comment (epage) [inline src/cargo/sources/git/utils.rs:182]: I think this change should be moved into `update_submodules`.  We should probably just mimic #10717 which also tells the user what is happening

That should probably provide good test examples

**Comment by brgl** [comment]
I've started working on a test-case that should more or less recreate what we do in yocto and started hitting a different issue - cargo bails out before even trying to update or checkout git revisions in offline mode like [here](https://github.com/rust-lang/cargo/blob/master/src/cargo/sources/git/source.rs#L167) or [here](https://github.com/rust-lang/cargo/blob/85e457e158db216a2938d51bc3b617a5a7fe6015/src/cargo/sources/git/utils.rs#L778). That doesn't seem correct. What if the remote we're using is a local one, with the `file://` URL? We don't need network to pull from it. Is there a specific reason for that? IMO these checks should be removed and we should let git fail on its own if the remote points outside the local environment.

**Comment by ehuss** [comment]
It's still not clear to me what this is trying to accomplish, so I can't really provide any advice. Before using `--offline`, one needs to run `cargo fetch` to download everything, which should fetch the submodules as well. Then, when using `--offline`, it shouldn't update the submodules as they are already downloaded.

**Comment by brgl** [comment]
> It's still not clear to me what this is trying to accomplish, so I can't really provide any advice. Before using `--offline`, one needs to run `cargo fetch` to download everything, which should fetch the submodules as well. Then, when using `--offline`, it shouldn't update the submodules as they are already downloaded.

I'm not sure how familiar you are with build tools aimed at embedded linux systems but the short story is this: yocto is a set of tools for creating custom linux distros. It uses a language called bitbake for creating so called recipes - they contain instructions on how to build various packages (different languages, build systems etc.).

For various reasons that I won't bore you with, yocto doesn't use `cargo fetch`. We cannot let a third party program download sources (even if they come from the same place - like crates.io) as we may want to use download mirrors, want to strictly validate the downloads etc. and we do it ourselves using various [fetcher classes](https://git.openembedded.org/bitbake/tree/lib/bb/fetch2).

So it prepares the database as if it were `cargo fetch` [here](https://git.openembedded.org/bitbake/tree/lib/bb/fetch2/crate.py) but the way the database is constructed almost never happens in regular cargo and results in taking this branch [here](https://github.com/rust-lang/cargo/blob/master/src/cargo/sources/git/source.rs#L148) and then failing [here](https://github.com/rust-lang/cargo/blob/master/src/cargo/sources/git/source.rs#L201).

I simply don't know cargo enough to recreate this situation. I can easily reproduce the issue in yocto and can give you instructions on how to do this, but you're probably not interested in non-standard usage which I can understand but I also want to fix the use-case that's making me carry out-of-tree openembedded-core patches.

**Comment by ehuss** [comment]
It sounds like this is trying to address an issue with an external tool trying to modify cargo's internal cache. I would recommend not doing that, as the internal cache isn't something that external tools should be modifying. There are other mechanisms for using sources fetched from external tools, such as [source replacement](https://doc.rust-lang.org/cargo/reference/source-replacement.html). 

I'm going to close as it doesn't appear to be a specific issue with using `--offline` with a properly fetched git dependency. If there is an issue with how offline works, it would probably be better to start with a description of the issue in #10727 focused on using cargo directly (and not modifying the cache directly).

If there are higher-level concerns about how to separately download sources that source replacement or other mechanisms don't address, then that might be good to discuss as a feature request or on the users or interals forum.

**Pull request closed without merging**